### PR TITLE
[SPARK-19605][DStream] Fail it if existing resource is not enough to run streaming job

### PR DESCRIPTION
**Pull request by uncleGen**
## What changes were proposed in this pull request?

For more detailed discussion, please review:
- [SPARK-19605](https://issues.apache.org/jira/browse/SPARK-19605)

## How was this patch tested?

add new unit test.

### CODE REVIEW

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 1]
```diff
@@ -437,6 +438,74 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
...
+    }
+  }
+
+  private def getTargetExecutorNumber(): Int = {
+    if (Utils.isDynamicAllocationEnabled(ssc.conf)) {
+      ssc.conf.get(DYN_ALLOCATION_MAX_EXECUTORS)
+    } else {
+      val targetNumExecutors =
+        sys.env.get("SPARK_EXECUTOR_INSTANCES").map(_.toInt).getOrElse(2)
+      // System property can override environment variable.
```
Comment (uncleGen) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: here "2" refers to "YarnSparkHadoopUtil.DEFAULT_NUMBER_EXECUTORS"

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala [hunk 2]
```diff
@@ -215,10 +215,6 @@ package object config {
 
   /* Executor configuration. */
 
-  private[spark] val EXECUTOR_CORES = ConfigBuilder("spark.executor.cores")
-    .intConf
-    .createWithDefault(1)
-
   private[spark] val EXECUTOR_MEMORY_OVERHEAD = ConfigBuilder("spark.yarn.executor.memoryOverhead")
```
Comment (uncleGen) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala]: multi-definition error in ApplicationMaster.scala, remove this as we add it in core

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 3]
```diff
@@ -938,7 +958,7 @@ object SlowTestReceiver {
 /** Streaming application for testing DStream and RDD creation sites */
 package object testPackage extends Assertions {
   def test() {
-    val conf = new SparkConf().setMaster("local").setAppName("CreationSite test")
+    val conf = new SparkConf().setMaster("local[2]").setAppName("CreationSite test")
     val ssc = new StreamingContext(conf, Milliseconds(100))
```
Comment (uncleGen) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: unit test fail here

**Comment by srowen** [comment]
(There is no detail in the JIRA that you point to.) I think we already discussed this in part? this duplicates some logic about how cluster resource management works. Does it really not work if not enough receivers can schedule? that seems like a condition to check somewhere else, like, log a warning that a batch can't operate or can't start because not all receivers scheduled. This could happen even if enough resources are theoretically available, just not available to the app.

**Comment by uncleGen** [comment]
@srowen 
> Does it really not work if not enough receivers can schedule? 

That's not what I want to express. What I mean is the stream output can not operate.

**Comment by srowen** [comment]
Hm it just seems like the wrong approach, to externally estimate whether in theory it won't schedule. It is certainly a problem if streaming doesn't work though users would already realize it. The error check or message could be more explicit but it seems like something the streaming machinery should know and warn about?

**Comment by uncleGen** [comment]
Let us call @zsxwing for some suggestions.

**Comment by zsxwing** [comment]
I agreed with @srowen. Streaming should not need to know the details about the mode, and when someone changes other modules, they won't know these codes inside streaming. I would like to see a cleaner solution.

In addition, when there are not enough resources in a cluster, even if the user sets a large core number, they will still not be able to run the streaming job. Hence, they should always be aware of this limitation of Streaming receivers, and it seems not worth to fix this issue.

**Pull request closed without merging**